harness/determined/pytorch/samplers.py

```diff
@@ -83,14 +89,11 @@ class DistributedBatchSampler(torch.utils.data.BatchSampler):
     DistributedSampler expects to bbe called before the BatchSampler, and
     additionally the DistributedSampler is meant to be a stand-alone sampler.
 
-    DistributedBatchSampler has the potential gotcha that when wrapping a
-    non-repeating BatchSampler, if the length of the BatchSampler is not
-    divisible by the number of replicas the length of the resulting
-    DistributedBatchSampler will differ based on the rank. In that case, the
-    divergent paths of multiple workers could cause problems during training.
-    PyTorchTrial always uses RepeatBatchSampler during training, PyTorchTrial
-    does not require that the workers stay in step during validation, so this
-    potential gotcha is not a problem in Determined.
+    DistributedBatchSampler reports equivalent len values for every worker, regardless of rank.
```

Comment:
Can we have the chief broadcast how long its shard is to the workers, and the workers use that, instead of having the dataset lie about its length?

I think, when we do support true-epoch-based training, this would be an undesirable behavior, and if you made it configurable, it would be difficult to document.  On the other hand, "in non-epoch based training, everybody gets the on_epoch_end call when the chief's epoch ends" feels more intuitive to explain, although it's a bit more complex to implement.